pkg/util/vsphere/vm/cdrom.go

```diff
@@ -0,0 +1,217 @@
+// Copyright (c) 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package vm
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/vmware/govmomi/object"
+	vimtypes "github.com/vmware/govmomi/vim25/types"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	imgregv1a1 "github.com/vmware-tanzu/image-registry-operator-api/api/v1alpha1"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha3"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+)
+
+const (
+	vmiKind  = "VirtualMachineImage"
+	cvmiKind = "Cluster" + vmiKind
+
+	serverIDSuffix = "?serverId="
+)
+
+// UpdateCdromDeviceChanges updates the CD-ROM devices in the VM ConfigSpec to
+// reflect the desired state in the K8s VM Spec.
+func UpdateCdromDeviceChanges(
+	vmCtx pkgctx.VirtualMachineContext,
+	client ctrlclient.Client,
+	curDevices object.VirtualDeviceList) ([]vimtypes.BaseVirtualDeviceConfigSpec, error) {
+	if vmCtx.MoVM.Summary.Runtime.PowerState == vimtypes.VirtualMachinePowerStatePoweredOn {
```

Comment:
This function should not check the power state. Or rather, I would like to see this function decomposed into one that does not care about the Kubernetes objects at all. For example:

```golang
// EnsureAllCDROMsHaveControllers ensures that all CD-ROM devices
// in the provided ConfigSpec point to a controller. If no controller
// exists, IDE and SATA controllers are added to the ConfigSpec as
// necessary for the CD-ROM devices.
//
// Please note the following table for the number of controllers of each type
// that are supported as well as how many devices (per controller) each
// supports:
//
// IDE
//   - controllers                                    2
//   - devices                                        2
//
// SATA
//   - controllers                                    4
//   - devices                                       30
func EnsureAllCDROMsHaveControllers(
     configSpec *vimtypes.ConfigSpec,
     existingDevices ...vimtypes.BaseVirtualDevice) error {

}